Nuestro programa al igual que el anterior, calcula las raices por el método de Newton Raphson (fuction NR) de la siguiente ecuación:

$$x \tan(x)-\sqrt{A-x^2}=y$$

analogamente nos interesa el valor de $y=0$, sin embargo ahora hemos definido el comando para que pueda evaluar errores y para esto necesitamos una cota del error.

Según Wikipedia, si definimos el error como:
$$Error=\frac{|x_{k+1}-x_{k}|}{|x_k|}$$
Entonces la cota, que en nuestra función esta definida como $E$, la función logra duplicar el número de dígitos que están bien aproximados cada vez que se le multiplique $*10^{-1}$, es por esto que exigimos que $E<1$

In [79]:
function NR(x,A,E)
    Error=1
    while Error>E
        y=x-(x*tan(x)-sqrt(complex(A-x^2)))/(x/(sqrt(complex(A-x^2)))+tan(x)+x*(sec(x))^2)
        Error=abs(y-x)/abs(y)
        x=y
    end
    return x
end

NR (generic function with 1 method)

Mostramos ahora, con un punto para hacer la recurción $x=.1$ y un valor fijo de $A=10$, que la aproximación puede llegar a variar mucho segun la cota que escribamos.

In [80]:
NR(.1,10,.1)

13.204407835032566 - 0.22043996609286307im

In [81]:
NR(.1,10,.01)

3.1650260633697456 - 0.007065117403015998im

In [63]:
NR(.1,10,.001)

3.160477511890897 + 0.00019257635128347352im

In [64]:
NR(.1,10,.0001)

3.161672009930193 - 2.031037064152365e-5im

In [65]:
NR(.1,10,.00001)

3.161642095037688 - 6.001141502407042e-11im

In [66]:
NR(.1,10,.000001)

3.161642095037688 - 6.001141502407042e-11im

In [67]:
NR(.1,10,.0000001)

3.1616420951060653 + 3.0355693173712014e-18im

In [68]:
NR(.1,10,.00000001)

3.1616420951060653 + 3.0355693173712014e-18im

In [69]:
NR(.1,10,.000000000000000000000000000000000000000001)

3.1616420951060653 - 4.555247183394356e-58im

In [70]:
NR(.1,10,.0000000000000000000000000000000000000000000000000000000000000000000000000000000000000001)

3.1616420951060653 - 3.638393417975715e-111im

Es así como vemos que la parte imaginaria tiende a cero, y en cambio la real se estanca, dandonos cuenta que 3.1616420951060653... es, en efecto una raiz de nuestro problema